File: embrace-android-okhttp3/src/main/java/io/embrace/android/embracesdk/okhttp3/EmbraceOkHttp3ApplicationInterceptor.java
package io.embrace.android.embracesdk.okhttp3;

import static io.embrace.android.embracesdk.config.behavior.NetworkSpanForwardingBehavior.TRACEPARENT_HEADER_NAME;
import static io.embrace.android.embracesdk.internal.utils.ThrowableUtilsKt.causeMessage;
import static io.embrace.android.embracesdk.internal.utils.ThrowableUtilsKt.causeName;

import java.io.IOException;

import io.embrace.android.embracesdk.Embrace;
import io.embrace.android.embracesdk.InternalApi;
import io.embrace.android.embracesdk.network.EmbraceNetworkRequest;
import io.embrace.android.embracesdk.network.http.EmbraceHttpPathOverride;
import io.embrace.android.embracesdk.network.http.HttpMethod;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * This interceptor will only intercept errors that client app experiences.
 * <p>
 * We used OkHttp3 application interceptor in this case because this interceptor
 * will be added first in the OkHttp3 interceptors stack. This allows us to catch network errors.
 * OkHttp3 network interceptors are added almost at the end of stack, they are closer to "Wire"
 * so they are not able to see network errors.
 * <p>
 * Application interceptors: - Don't need to worry about intermediate responses like
 * redirects and retries. - Are always invoked once, even if the HTTP response is served
 * from the cache. - Observe the application's original intent. Unconcerned with OkHttp-injected
 * headers like If-None-Match. - Permitted to short-circuit and not call
 * Chain.proceed(). - Permitted to retry and make multiple calls to Chain.proceed().
 * <p>
 * We used the EmbraceGraphQLException to capture the custom path added in the intercept
 * chain process for client errors on graphql requests.
 */
@InternalApi
public class EmbraceOkHttp3ApplicationInterceptor implements Interceptor {
    static final String UNKNOWN_EXCEPTION = "Unknown";
    static final String UNKNOWN_MESSAGE = "An error occurred during the execution of this network request";
    final Embrace embrace;

    private final SdkFacade sdkFacade;

    public EmbraceOkHttp3ApplicationInterceptor() {
        this(Embrace.getInstance(), new SdkFacade());
    }

    EmbraceOkHttp3ApplicationInterceptor(Embrace embrace, SdkFacade sdkFacade) {
        this.embrace = embrace;
        this.sdkFacade = sdkFacade;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        long startTime = System.currentTimeMillis();
        Request request = chain.request();
        try {
            // we are not interested in response, just proceed
            return chain.proceed(request);
        } catch (EmbraceCustomPathException e) {
            if (embrace.isStarted()) {
                String urlString = EmbraceHttpPathOverride.getURLString(new EmbraceOkHttp3PathOverrideRequest(request), e.getCustomPath());

                embrace.recordNetworkRequest(
                    EmbraceNetworkRequest.fromIncompleteRequest(
                        urlString,
                        HttpMethod.fromString(request.method()),
                        startTime,
                        System.currentTimeMillis(),
                        causeName(e, UNKNOWN_EXCEPTION),
                        causeMessage(e, UNKNOWN_MESSAGE),
                        request.header(embrace.getTraceIdHeader()),
                        sdkFacade.isNetworkSpanForwardingEnabled() ? request.header(TRACEPARENT_HEADER_NAME) : null,
                        null
                    )
                );
            }
            throw e;
        } catch (Exception e) {
            // we are interested in errors.
            if (embrace.isStarted()) {
                String urlString = EmbraceHttpPathOverride.getURLString(new EmbraceOkHttp3PathOverrideRequest(request));
                String errorType = e.getClass().getCanonicalName();
                String errorMessage = e.getMessage();

                embrace.recordNetworkRequest(
                    EmbraceNetworkRequest.fromIncompleteRequest(
                        urlString,
                        HttpMethod.fromString(request.method()),
                        startTime,
                        System.currentTimeMillis(),
                        errorType != null ? errorType : UNKNOWN_EXCEPTION,
                        errorMessage != null ? errorMessage : UNKNOWN_MESSAGE,
                        request.header(embrace.getTraceIdHeader()),
                        sdkFacade.isNetworkSpanForwardingEnabled() ? request.header(TRACEPARENT_HEADER_NAME) : null,
                        null
                    )
                );
            }
            throw e;
        }
    }
}

File: embrace-android-okhttp3/src/main/java/io/embrace/android/embracesdk/okhttp3/EmbraceOkHttp3ApplicationInterceptor.kt
package io.embrace.android.embracesdk.okhttp3

import io.embrace.android.embracesdk.Embrace
import io.embrace.android.embracesdk.internal.EmbraceInternalApi
import io.embrace.android.embracesdk.internal.EmbraceInternalApi.Companion.CUSTOM_TRACE_ID_HEADER_NAME
import io.embrace.android.embracesdk.internal.network.http.EmbraceHttpPathOverride
import io.embrace.android.embracesdk.network.EmbraceNetworkRequest
import io.embrace.android.embracesdk.network.http.HttpMethod
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import java.io.IOException

/**
 * This interceptor will only intercept errors that client app experiences.
 *
 * We used OkHttp application interceptor in this case because this interceptor
 * will be added first in the OkHttp3 interceptors stack. This allows us to catch network errors.
 * OkHttp network interceptors are added almost at the end of stack, they are closer to "the wire"
 * so they are not able to see network errors.
 *
 * We used the [EmbraceCustomPathException] to capture the custom path added in the interceptor
 * chain process for client errors on requests to a generic URL like a GraphQL endpoint.
 */
class EmbraceOkHttp3ApplicationInterceptor(
    private val embrace: Embrace,
    private val embraceInternalApi: EmbraceInternalApi,
) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val startTime = embrace.getSdkCurrentTimeMs()
        val request: Request = chain.request()
        return try {
            // we are not interested in response, just proceed
            chain.proceed(request)
        } catch (e: EmbraceCustomPathException) {
            if (embrace.isStarted) {
                val urlString =
                    EmbraceHttpPathOverride.getURLString(EmbraceOkHttp3PathOverrideRequest(request), e.customPath)
                embrace.recordNetworkRequest(
                    EmbraceNetworkRequest.fromIncompleteRequest(
                        urlString,
                        HttpMethod.fromString(request.method),
                        startTime,
                        embrace.getSdkCurrentTimeMs(),
                        causeName(e, UNKNOWN_EXCEPTION),
                        causeMessage(e, UNKNOWN_MESSAGE),
                        request.header(CUSTOM_TRACE_ID_HEADER_NAME),
                        if (embraceInternalApi.internalInterface.isNetworkSpanForwardingEnabled()) {
                            request.header(
                                TRACEPARENT_HEADER_NAME
                            )
                        } else {
                            null
                        },
                        null
                    )
                )
            }
            throw e
        } catch (e: Exception) {
            // we are interested in errors.
            if (embrace.isStarted) {
                val urlString = EmbraceHttpPathOverride.getURLString(EmbraceOkHttp3PathOverrideRequest(request))
                val errorType = e.javaClass.canonicalName
                val errorMessage = e.message
                embrace.recordNetworkRequest(
                    EmbraceNetworkRequest.fromIncompleteRequest(
                        urlString,
                        HttpMethod.fromString(request.method),
                        startTime,
                        embrace.getSdkCurrentTimeMs(),
                        errorType ?: UNKNOWN_EXCEPTION,
                        errorMessage ?: UNKNOWN_MESSAGE,
                        request.header(CUSTOM_TRACE_ID_HEADER_NAME),
                        if (embraceInternalApi.internalInterface.isNetworkSpanForwardingEnabled()) {
                            request.header(
                                TRACEPARENT_HEADER_NAME
                            )
                        } else {
                            null
                        },
                        null
                    )
                )
            }
            throw e
        }
    }

    internal companion object {
        internal const val TRACEPARENT_HEADER_NAME = "traceparent"
        internal const val UNKNOWN_EXCEPTION = "Unknown"
        internal const val UNKNOWN_MESSAGE = "An error occurred during the execution of this network request"

        /**
         * Return the canonical name of the cause of a [Throwable]. Handles null elements throughout,
         * including the throwable and its cause, in which case [defaultName] is returned
         */
        internal fun causeName(throwable: Throwable?, defaultName: String = ""): String {
            return throwable?.cause?.javaClass?.canonicalName ?: defaultName
        }

        /**
         * Return the message of the cause of a [Throwable]. Handles null elements throughout,
         * including the throwable and its cause, in which case [defaultMessage] is returned
         */
        internal fun causeMessage(throwable: Throwable?, defaultMessage: String = ""): String {
            return throwable?.cause?.message ?: defaultMessage
        }
    }
}
